File: src/main/java/main/utils/database/mongodb/cache/redis/RedisCache.java
package main.utils.database.mongodb.cache.redis;

import com.mongodb.client.MongoCollection;
import main.constants.ENV;
import main.main.Config;
import main.utils.json.AbstractJSON;
import org.bson.Document;
import org.bson.types.ObjectId;
import org.json.JSONArray;
import org.json.JSONObject;
import redis.clients.jedis.JedisPooled;

import java.util.HashMap;
import java.util.Map;

public abstract class RedisCache implements AbstractJSON {

    private final String cacheID;
    private final JedisPooled jedis;

    protected RedisCache(String cacheID) {
        this.cacheID = cacheID + "#"+ Config.get(ENV.MONGO_DATABASE_NAME) +"#";
        this.jedis = RedisDB.getInstance().getJedis();
    }

    void hsetJSON(String identifier, HashMap<String, JSONObject> hash) {
        final HashMap<String, String> newHash = new HashMap<>();
        hash.forEach((key, val) -> newHash.put(key, val.toString()));
        hset(identifier, newHash);
    }

    void hset(String identifier, HashMap<String, String> hash) {
        jedis.hset(cacheID + identifier, hash);
    }

    String hget(String identifier, String key) {
        return jedis.hget(cacheID + identifier, key);
    }

    JSONObject hgetJSON(String identifier, String key) {
        return new JSONObject(hget(identifier, key));
    }

    Map<String, String> hgetAll(String key) {
        return jedis.hgetAll(key);
    }

    void setex(String identifier, int seconds, String value) {
        jedis.setex(cacheID + identifier, seconds, value);
    }

    void setex(String identifier, int seconds, JSONObject value) {
        setex(identifier, seconds, value.toString());
    }

    void setex(long identifier, int seconds, JSONObject value) {
        setex(String.valueOf(identifier), seconds, value.toString());
    }

    String get(String identifier) {
        return jedis.get(cacheID + identifier);
    }

    String get(long identifier) {
        return get(String.valueOf(identifier));
    }

    void del(String identifier) {
        jedis.del(cacheID + identifier);
    }

    void del(long identifier) {
        del(String.valueOf(identifier));
    }

    public void updateCache(String identifier, Document document) {
        jedis.del(identifier);
        jedis.set(cacheID + identifier, document.toJson());
    }

    public void updateCache(String identifier, Document document, int expiration) {
        jedis.del(identifier);
        jedis.setex(cacheID + identifier, expiration, document.toJson());
    }

    public void updateCache(String identifier, JSONObject object) {
        jedis.del(identifier);
        jedis.set(cacheID + identifier, object.toString());
    }

    public void updateCache(String identifier, JSONObject object, int expiration) {
        jedis.del(identifier);
        jedis.setex(cacheID + identifier, expiration, object.toString());
    }

    public void updateCacheObjects(HashMap<String, JSONObject> objects) {
        final HashMap<String, Document> documents = new HashMap<>();
        objects.forEach((key, object) -> documents.put(key, Document.parse(object.toString())));
        updateCache(documents);
    }

    public void updateCache(HashMap<String, Document> documents) {

        for (var document : documents.entrySet()) {
            jedis.del(document.getKey());
            jedis.setex(cacheID + document.getKey(), 3600, document.getValue().toJson());
        }
    }

    public void removeFromCache(String id) {
        del(id);
    }

    public JSONObject getCacheJSON(String identifier) {
        return new JSONObject(get(identifier));
    }

    public JSONObject getJSON(String id) {
        String source = get(id);
        if (source == null)
            return null;
        return new JSONObject(source);
    }

    public JSONObject getJSONByGuild(String gid) {
        return getJSON(gid);
    }

    public JSONObject getJSONByGuild(long gid) {
        return getJSON(String.valueOf(gid));
    }

    private JSONObject collectionToJSON(MongoCollection<Document> collection) {
        JSONObject collectionObj = new JSONObject();
        JSONArray documentArr = new JSONArray();

        collection.find().forEach(doc -> documentArr.put(new JSONObject(doc.toJson())));

        collectionObj.put(DatabaseRedisCache.CacheField.DOCUMENTS.toString(), documentArr);
        return collectionObj;
    }

    public JSONObject getCache(String id) {
        return new JSONObject(get(cacheID + id));
    }

    boolean arrayHasObject(JSONArray array, ObjectId object) {
        for (int i = 0; i < array.length(); i++) {
            if (array.getJSONObject(i).getJSONObject("_id").getString("$oid").equals(object.toString()))
                return true;
        }
        return false;
    }

    int getIndexOfObjectInArray(JSONArray array, ObjectId object) {
        if (!arrayHasObject(array, object))
            throw new NullPointerException("There was no such object found in the array!");

        for (int i = 0; i < array.length(); i++)
            if (array.getJSONObject(i).getJSONObject("_id").getString("$oid").equals(object.toString()))
                return i;
        return -1;
    }
}

File: src/main/kotlin/main/utils/database/mongodb/cache/redis/RedisCache.kt
package main.utils.database.mongodb.cache.redis

import com.mongodb.client.MongoCollection
import kotlinx.coroutines.coroutineScope
import main.main.Config
import org.bson.Document
import org.json.JSONArray
import org.json.JSONObject
import java.util.function.Consumer

abstract class RedisCache protected constructor(cacheID: String) {

    companion object {
        private val jedis = RedisDB.jedis
    }

    private val cacheID = "$cacheID#${Config.MONGO_DATABASE_NAME}#"

    protected fun hsetJSON(identifier: String, hash: HashMap<String, JSONObject>) {
        val newHash = HashMap<String, String>()
        hash.forEach { (key: String, `val`: JSONObject) ->
            newHash[key] = `val`.toString()
        }

        hset(identifier, newHash)
    }

    protected fun hset(identifier: String, hash: HashMap<String, String>) {
        jedis.hset(cacheID + identifier, hash)
    }

    protected fun hset(hash: HashMap<String, String>) {
        hset(cacheID, hash)
    }

    protected fun hget(identifier: String, key: String): String? {
        return jedis.hget(cacheID + identifier, key)
    }

    protected fun hgetJSON(identifier: String, key: String): JSONObject {
        return JSONObject(hget(identifier, key))
    }

    protected fun hgetAll(key: String): Map<String, String> {
        return jedis.hgetAll(key)
    }

    protected fun setex(identifier: String, seconds: Long, value: String): String {
        return jedis.setex(cacheID + identifier, seconds, value)
    }

    protected open fun set(identifier: String, value: String) {
        jedis.set(cacheID + identifier, value)
    }

    protected fun set(value: String): String {
        return jedis.set(cacheID, value)
    }

    protected fun setex(identifier: String, seconds: Long, value: JSONObject) {
        setex(identifier, seconds, value.toString())
    }

    protected fun setex(identifier: Long, seconds: Long, value: JSONObject) {
        setex(identifier.toString(), seconds, value.toString())
    }

    protected open fun get(identifier: String): String? {
        val item = jedis.get(cacheID + identifier)
        return if (item == "nil") null else item
    }

    protected open fun get(identifier: Long): String? {
        return get(identifier.toString())
    }

    protected fun get(): String? {
        return jedis.get(cacheID)
    }

    protected fun del(identifier: String): Long {
        return jedis.del(cacheID + identifier)
    }

    protected fun del(identifier: Long): Long {
        return del(identifier.toString())
    }

    protected fun del(): Long {
        return jedis.del(cacheID)
    }

    protected fun exists(): Boolean {
        return jedis.exists(cacheID)
    }

    protected fun exists(identifier: String): Boolean {
        return jedis.exists(cacheID + identifier)
    }

    open fun updateCache(identifier: String, document: Document) {
        set(cacheID + identifier, document.toJson())
    }

    fun updateCache(identifier: String, document: Document, expiration: Long) {
        setex(cacheID + identifier, expiration, document.toJson())
    }

    open fun updateCache(identifier: String, `object`: JSONObject) {
        set(cacheID + identifier, `object`.toString())
    }

    fun updateCache(identifier: String, `object`: JSONObject, expiration: Long) {
        setex(cacheID + identifier, expiration, `object`.toString())
    }

    open fun updateCacheObjects(objects: HashMap<String, JSONObject>) {
        val documents = HashMap<String, Document>()
        objects.forEach { (key: String, `object`: JSONObject) ->
            documents[key] = Document.parse(`object`.toString())
        }
        updateCache(documents)
    }

    open fun updateCache(documents: HashMap<String, Document>) {
        for ((key, value) in documents)
            jedis.setex(cacheID + key, 3600, value.toJson())
    }

    fun removeFromCache(id: String) {
        del(id)
    }

    fun getCacheJSON(identifier: String): JSONObject {
        return JSONObject(get(identifier))
    }

    open fun getJSON(id: String): JSONObject? {
        val source = get(id) ?: return null
        return JSONObject(source)
    }

    open fun getJSONByGuild(gid: String): JSONObject? {
        return getJSON(gid)
    }

    fun getJSONByGuild(gid: Long): JSONObject? {
        return getJSON(gid.toString())
    }

    private fun collectionToJSON(collection: MongoCollection<Document>): JSONObject {
        val collectionObj = JSONObject()
        val documentArr = JSONArray()
        collection.find().forEach(Consumer { doc: Document ->
            documentArr.put(
                JSONObject(doc.toJson())
            )
        })
        collectionObj.put(DatabaseRedisCache.CacheField.DOCUMENTS.toString(), documentArr)
        return collectionObj
    }

    open fun getCache(id: String): JSONObject {
        return JSONObject(get(cacheID + id))
    }

}